projects/optic/src/commands/capture/capture.ts

```diff
@@ -128,187 +155,117 @@ const getCaptureAction =
-      }
+    // start proxy
+    const proxy = new ProxyInstance(
+      options.serverOverride || captureConfig.server.url
+    );
+    await proxy.start(
+      options.proxyPort ? Number(options.proxyPort) : undefined
     );
+
+    // parse optic.yml
     let spec = await loadSpec(filePath, config, {
       strict: false,
       denormalize: false,
     });
 
-    // start the app
-    let app: ChildProcessWithoutNullStreams | undefined = undefined;
+    // start app
+    let app: ChildProcessWithoutNullStreams | undefined;
     if (!options.serverOverride && captureConfig.server.command) {
-      const cmd = commandSplitter(captureConfig.server.command);
-      app = spawn(cmd.cmd, cmd.args, { detached: true });
-
-      app.stderr.on('data', (data) => {
-        logger.error(data.toString());
-      });
-    }
-
-    // wait until the server is ready
-    const readyInterval = captureConfig.server.ready_interval || 1000;
-
-    // since ready_endpoint is not required always wait one interval. without ready_endpoint,
-    // ready_interval must be at least the time it takes to start the server.
-    await wait(readyInterval);
-
-    if (captureConfig.server.ready_endpoint) {
-      const readyUrl = urljoin(serverUrl, captureConfig.server.ready_endpoint);
-      const readyTimeout = captureConfig.server.ready_timeout || 3 * 60 * 1_000; // 3 minutes
-
-      const now = Date.now();
-      const spinner = ora('Waiting for server to come online...');
-      spinner.start();
-      spinner.color = 'blue';
-
-      const checkServer = (): Promise<boolean> =>
-        fetch(readyUrl)
-          .then((res) => String(res.status).startsWith('2'))
-          .catch((e) => {
-            logger.debug(e);
-            return false;
-          });
-
-      let done = false;
-      while (!done) {
-        const isReady = await checkServer();
-        if (isReady) {
-          spinner.succeed('Server check passed');
-          done = true;
-        } else if (Date.now() > now + readyTimeout) {
-          throw new UserError('Server check timed out.');
-        }
-        await wait(readyInterval);
-      }
+      const serverDir =
+        captureConfig.server.dir === undefined
+          ? config.root
+          : captureConfig.server.dir;
+      const timeout = captureConfig.server.ready_timeout || 3 * 60 * 1_000; // 3 minutes
+      const readyInterval = captureConfig.server.ready_interval || 1000;
+
+      app = await startApp(
+        captureConfig.server.command,
+        serverDir,
+        captureConfig.server.ready_endpoint,
+        readyInterval,
+        timeout,
+        proxy.targetUrl
+      );
     }
 
     // make requests
-    const captures = new GroupedCaptures(
-      trafficDirectory,
-      specToOperations(spec.jsonLike).flatMap((o) =>
-        o.methods.map((m) => ({ method: m, path: o.pathPattern }))
-      )
-    );
-
     let errors: any[] = [];
     try {
```

Comment:
`app = await startApp` should be included in this try catch block. Otherwise any errors thrown while starting the app will not be caught and will not tear down the proxy server

Reply:
wrapped it in its own try-catch, https://github.com/opticdev/optic/pull/2097/commits/ac9fe617551983793c3c9f05d85044b4e0bb334a—otherwise there's bigger changes to make since we don't always run the server.

also now throwing an error when the server exits with non-zero code. doesn't quite handle all early termination situations (like an early exit with 0) but we're expecting the server to behave reasonably to a certain extent.